File: org.summer.view.widget/src/org/summer/view/widget/CollectionChangedEventManager.java
package org.summer.view.widget;

import org.summer.view.widget.collection.NotifyCollectionChangedEventArgs;
 /// <summary>
    /// Manager for the INotifyCollectionChanged.CollectionChanged event.
    /// </summary>
    public class CollectionChangedEventManager extends WeakEventManager
    {
//        #region Constructors
    	//
        //  Constructors
        //
 
        private CollectionChangedEventManager()
        {
        }
 
//        #endregion Constructors
// 
//        #region Public Methods
  
        //
        //  Public Methods
        //
 
        /// <summary>
        /// Add a listener to the given source's event.
        /// </summary>
        public static void AddListener(INotifyCollectionChanged source, IWeakEventListener listener)
        {
            CurrentManager.ProtectedAddListener(source, listener);
        }
 
        /// <summary>
        /// Remove a listener to the given source's event.
        /// </summary>
        public static void RemoveListener(INotifyCollectionChanged source, IWeakEventListener listener)
        {
            CurrentManager.ProtectedRemoveListener(source, listener);
        }
  
//        #endregion Public Methods
// 
//        #region Protected Methods
  
        //
        //  Protected Methods
        //
 
        /// <summary>
        /// Listen to the given source for the event.
        /// </summary>
        protected void StartListening(Object source)
        {
            INotifyCollectionChanged typedSource = (INotifyCollectionChanged)source;
            typedSource.CollectionChanged += new NotifyCollectionChangedEventHandler(OnCollectionChanged);
        }
 
        /// <summary>
        /// Stop listening to the given source for the event.
        /// </summary>
        protected void StopListening(Object source)
        {
            INotifyCollectionChanged typedSource = (INotifyCollectionChanged)source;
            typedSource.CollectionChanged -= new NotifyCollectionChangedEventHandler(OnCollectionChanged);
        }
 
//        #endregion Protected Methods
// 
//        #region Private Properties
 
        //
        //  Private Properties
        //
  
        // get the event manager for the current thread
        private static CollectionChangedEventManager CurrentManager
        {
            get
            {
                Type managerType = typeof(CollectionChangedEventManager);
                CollectionChangedEventManager manager = (CollectionChangedEventManager)GetCurrentManager(managerType);
  
                // at first use, create and register a new manager
                if (manager == null)
                {
                    manager = new CollectionChangedEventManager();
                    SetCurrentManager(managerType, manager);
                }
  
                return manager;
            }
        }
 
//        #endregion Private Properties
// 
//        #region Private Methods
 
        //
        //  Private Methods
        //
  
        // event handler for CollectionChanged event
        private void OnCollectionChanged(Object sender, NotifyCollectionChangedEventArgs args)
        {
            DeliverEvent(sender, args);
        }
  
//        #endregion Private Methods
    }

File: summer/WebContent/js/system/collections/specialized/CollectionChangedEventManager.js
/**
 * CollectionChangedEventManager
 */

define(["dojo/_base/declare", "system/Type"], function(declare, Type){
	var CollectionChangedEventManager = declare("CollectionChangedEventManager", null,{
        /// <summary> 
        /// Return a new list to hold listeners to the event.
        /// </summary> 
//        protected override ListenerList 
        NewListenerList:function()
        {
            return new ListenerList/*<NotifyCollectionChangedEventArgs>*/();
        }, 

        /// <summary> 
        /// Listen to the given source for the event. 
        /// </summary>
//        protected override void 
        StartListening:function(/*object*/ source) 
        {
            /*INotifyCollectionChanged*/var typedSource = source;
            typedSource.CollectionChanged += new NotifyCollectionChangedEventHandler(OnCollectionChanged);
        }, 

        /// <summary> 
        /// Stop listening to the given source for the event. 
        /// </summary>
//        protected override void 
        StopListening:function(/*object*/ source) 
        {
            /*INotifyCollectionChanged*/var typedSource = source;
            typedSource.CollectionChanged -= new NotifyCollectionChangedEventHandler(OnCollectionChanged);
        },
 
        // event handler for CollectionChanged event
//        private void 
        OnCollectionChanged:function(/*object*/ sender, /*NotifyCollectionChangedEventArgs*/ args) 
        { 
            DeliverEvent(sender, args);
        } 
	});
	
	Object.defineProperties(CollectionChangedEventManager.prototype,{
	});
	
	Object.defineProperties(CollectionChangedEventManager,{

        // get the event manager for the current thread
//        private static CollectionChangedEventManager 
        CurrentManager: 
        { 
            get:function()
            { 
                /*Type*/var managerType = CollectionChangedEventManager.Type;
                /*CollectionChangedEventManager*/
                var manager = GetCurrentManager(managerType);

                // at first use, create and register a new manager 
                if (manager == null)
                { 
                    manager = new CollectionChangedEventManager(); 
                    SetCurrentManager(managerType, manager);
                } 

                return manager;
            }
        } 
	});
	/// <summary>
    /// Add a listener to the given source's event. 
    /// </summary>
//    public static void 
    CollectionChangedEventManager.AddListener = function(/*INotifyCollectionChanged*/ source, /*IWeakEventListener*/ listener) 
    { 
        if (source == null)
            throw new ArgumentNullException("source"); 
        if (listener == null)
            throw new ArgumentNullException("listener");

        CurrentManager.ProtectedAddListener(source, listener); 
    };

    /// <summary> 
    /// Remove a listener to the given source's event.
    /// </summary> 
//    public static void 
    CollectionChangedEventManager.RemoveListener = function(/*INotifyCollectionChanged*/ source, /*IWeakEventListener*/ listener)
    {
        /* for app-compat, allow RemoveListener(null, x) - it's a no-op (see Dev10 796788)
        if (source == null) 
            throw new ArgumentNullException("source");
        */ 
        if (listener == null) 
            throw new ArgumentNullException("listener");

        CurrentManager.ProtectedRemoveListener(source, listener);
    };

    /// <summary> 
    /// Add a handler for the given source's event.
    /// </summary> 
//    public static void 
    CollectionChangedEventManager.AddHandler = function(/*INotifyCollectionChanged*/ source, /*EventHandler<NotifyCollectionChangedEventArgs>*/ handler) 
    {
        if (handler == null) 
            throw new ArgumentNullException("handler");

        CurrentManager.ProtectedAddHandler(source, handler);
    };

    /// <summary> 
    /// Remove a handler for the given source's event. 
    /// </summary>
//    public static void 
    CollectionChangedEventManager.RemoveHandler = function(/*INotifyCollectionChanged*/ source, /*EventHandler<NotifyCollectionChangedEventArgs>*/ handler) 
    {
        if (handler == null)
            throw new ArgumentNullException("handler");

        CurrentManager.ProtectedRemoveHandler(source, handler);
    };
    
	CollectionChangedEventManager.Type = new Type("CollectionChangedEventManager", CollectionChangedEventManager, [Object.Type]);
	return CollectionChangedEventManager;
});

        

